selfservice/strategy/code/strategy_recovery.go

```diff
@@ -0,0 +1,609 @@
+	if _, err := s.deps.RecoveryCodePersister().CreateRecoveryCode(ctx, dto); err != nil {
+		s.deps.Writer().WriteError(w, r, err)
+		return
+	}
+
+	s.deps.Audit().
+		WithField("identity_id", id.ID).
+		WithSensitiveField("recovery_code", code).
+		Info("A recovery code has been created.")
+
+	body := &selfServiceRecoveryCode{
+		ExpiresAt: flow.ExpiresAt.UTC(),
+		RecoveryLink: urlx.CopyWithQuery(
+			s.deps.Config().SelfServiceFlowRecoveryUI(ctx),
+			url.Values{
+				"flow": {flow.ID.String()},
+			}).String(),
+		RecoveryCode: code,
+	}
+
+	s.deps.Writer().WriteCode(w, r, http.StatusCreated, body, herodot.UnescapedHTML)
+}
+
+// swagger:model submitSelfServiceRecoveryFlowWithCodeMethodBody
+// nolint:deadcode,unused
+type submitSelfServiceRecoveryFlowWithCodeMethodBody struct {
+	// Email to Recover
+	//
+	// Needs to be set when initiating the flow. If the email is a registered
+	// recovery email, a recovery link will be sent. If the email is not known,
+	// a email with details on what happened will be sent instead.
+	//
+	// format: email
+	// required: false
+	Email string `json:"email" form:"email"`
+
+	// Code from recovery email
+	//
+	// Sent to the user once a recovery has been initiated and is used to prove
+	// that the user is in possession of the email
+	//
+	// required: false
+	Code string `json:"code" form:"code"`
+
+	// Sending the anti-csrf token is only required for browser login flows.
+	CSRFToken string `form:"csrf_token" json:"csrf_token"`
+
+	// Method supports `link` and `code` only right now.
+	//
+	// required: true
+	Method string `json:"method"`
+}
+
+func (s Strategy) isCodeFlow(f *recovery.Flow) bool {
+	value, err := f.Active.Value()
+	if err != nil {
+		return false
+	}
+	return value == s.RecoveryNodeGroup().String()
+}
+
+func (s *Strategy) Recover(w http.ResponseWriter, r *http.Request, f *recovery.Flow) (err error) {
+	if !s.isCodeFlow(f) {
+		return errors.WithStack(flow.ErrStrategyNotResponsible)
+	}
+
+	body, err := s.decodeRecovery(r)
+	if err != nil {
+		return s.HandleRecoveryError(w, r, nil, body, err)
+	}
+	ctx := r.Context()
+
+	// If a CSRF violation occurs the flow is most likely FUBAR, as the user either lost the CSRF token, or an attack occured.
+	// In this case, we just issue a new flow and "abandon" the old flow.
+	if err := flow.EnsureCSRF(s.deps, r, f.Type, s.deps.Config().DisableAPIFlowEnforcement(ctx), s.deps.GenerateCSRFToken, body.CSRFToken); err != nil {
+		return s.retryRecoveryFlowWithError(w, r, flow.TypeBrowser, err)
+	}
+
+	sID := s.RecoveryStrategyID()
+
+	// If the email is present in the submission body, the user needs a new code via resend
+	if f.State != recovery.StateChooseMethod && len(body.Email) == 0 {
+		if err := flow.MethodEnabledAndAllowed(ctx, sID, sID, s.deps); err != nil {
+			return s.HandleRecoveryError(w, r, nil, body, err)
+		}
+
+		f.SubmitCount++
+		if err := s.deps.RecoveryFlowPersister().UpdateRecoveryFlow(ctx, f); err != nil {
```

Comment:
There were a lot of changes, and a lot of comments raised and addressed, so I'll just focus on this: Are you really sure that the `SubmitCount` is updated atomically? What we want to prevent is that multiple requests come in concurrently with different brute-forced codes, and then concurrently `SubmitCount` is set to `1` and the database is updated.

I would have expected some kind of database transaction beginning from when the `Recover` function starts to until we know whether the code matched.

You have a test "description=should not be able to use an invalid code more than 5 times", that tests that the count works, but it performs the checks concurrently.

In order to protect us from brute-forcing 8 digits, I'd really want to have the concurrent case covered. Am I missing something about the flow works? Still new to the Kratos code base.

Reply:
Yes, you're right. I moved the `submit_count > 5` check inside the transaction.

However, I am not sure how to test the concurrent submission issue, or rather, how we should handle concurrent submissions in general. I wrote a quick test case that executes the same recovery request concurrently, but that leads to `500 database table is locked: selfservice_recovery_flows`. I think this is only an issue on SQLite, as the others should have better (no) locking mechanisms. But I am no expert on this... We are running all tests against SQLite. Since you know the code base better, WDYT @aeneasr?.

Reply:
I am now looking at writing a test to reproduce this. There are several issues with "just spawn a bunch of goroutines that execute the same request". I have talked to @hperl, and we discussed a few of them, but these are still there, and I am not sure how to solve these:
1. The `NewErrorTestServer` reads errors from the DB, which fails with "database is locked". I assume this is due to the many requests, but can't verify this.
2. Some requests fail with 
```
Error:      	Received unexpected error:
Post "http://127.0.0.1:54193/recovery-ts?flow=12a1c322-dd4b-43ce-8da6-0870c2a9f4fd": EOF
```
I assume, this is due to the test servers being overloaded. 

At this point, I am not even sure if `strategy_recovery.go` is the correct place to do this. Do you know if we have done this type of testing before, or do you have experience with it? I could use some pointers here.

Reply:
Since we have logic in SQL in a transaction to solve this, I would be fine with not testing it and instead focus on getting this shipped. Maybe add some code comments to make sure that people understand why this is needed!